config/v3_4_experimental/types/config.go

```diff
@@ -25,3 +31,50 @@ var (
 		PreRelease: "experimental",
 	}
 )
+
+func (cfg Config) Validate(c path.ContextPath) (r report.Report) {
+	systemdPath := "/etc/systemd/system/"
+	units := make(map[string]string)
+
+	for _, unit := range cfg.Systemd.Units {
+		if unit.Dropins == nil {
+			units[unit.Name] = systemdPath + unit.Name
+		} else {
+			for _, dropin := range unit.Dropins {
+				units[dropin.Name] = systemdPath + unit.Name + ".d/" + dropin.Name
+			}
+		}
+	}
+	for i, f := range cfg.Storage.Files {
+		if validateSystemdConflict(f.Path, units) {
+			r.AddOnError(c.Append("storage", "files", i, "path"), errors.ErrFileConflictsSystemd)
+		}
+	}
+	for i, d := range cfg.Storage.Directories {
+		if validateSystemdConflict(d.Path, units) {
+			r.AddOnError(c.Append("storage", "directories", i, "path"), errors.ErrDirConflictsSystemd)
+		}
+	}
+	for i, l := range cfg.Storage.Links {
+		if validateSystemdConflict(l.Path, units) {
+			r.AddOnError(c.Append("storage", "links", i, "path"), errors.ErrLinkConflictsSystemd)
+		}
+	}
+	return
+}
+
+func validateSystemdConflict(p string, units map[string]string) bool {
+	systemdPath := "/etc/systemd/system/"
+
+	if strings.HasPrefix(p, systemdPath) {
+		parts := strings.Split(p, "/")
+		fileName := parts[len(parts)-1]
+		_, exists := units[fileName]
+		if exists {
+			if strings.Compare(p, units[fileName]) == 0 {
+				return true
+			}
+		}
```

Comment:
This is doing a lot of work just to look up a filename.  Is there any reason we can't make `units` just a set of full pathnames, and do a straight existence check here?